Use the plugin settings to choose PRB locations instead of filters

reykjalin opened the pull request:
## Changes proposed in this Pull Request:

Fixes #1640

Instead of relying on filters to configure where to display Payment Request Buttons we now use a `multiselect` element in the Stripe settings to choose the locations we want to show PRBs in.

- Removed the `wc_stripe_hide_payment_request_on_product_page` filter.
- Removed the `wc_stripe_show_payment_request_on_checkout` filter.
- Removed the `wc_stripe_show_payment_request_on_cart` filter.
- Added the `payment_request_button_locations` setting.
- Move the results of the removed filters into the plugin settings.

## Testing instructions

> **Note:** You'll need database access to properly test this feature. The testing instructions will assume you're using phpMyAdmin.

> **Note:** You may see deprecation warnings if you have `WP_DEBUG` set to `true` while the code snippets (added later) are enabled.

### Test default values

1. `git checkout fix/1640-add-prb-location-settings`.
2. Clear any use of the removed filters (mentioned above).
3. Change the `WC_STRIPE_VERSION` constant [in `woocommerce-gateway-stripe.php:24`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/38744bb1c4add59869f5067734c0960f726b1228/woocommerce-gateway-stripe.php#L24) from whatever it's current value is to any other version, e.g. change `5.3.0` to `5.4.0`.
4. Open **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**
5. Make sure the **Payment Request Button Locations** setting is set to **Cart** and **Product**.
6. Make sure the PRBs are loaded on the **Product** pages.
7. Make sure the PRBs are loaded on the **Cart** page (both Blocks and Shortcode versions).
8. Make sure the PRBs are not loaded on the **Checkout** page (both Blocks and Shortcode versions).
9. Open phpMyAdmin, navigate to the **wp_options** table, toggle **Show all** on, search for **woocommerce_stripe_settings**, and click **Edit**.

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/13835680/129649524-a174d450-e508-460a-9ab7-6b7514a0b0a0.png">

10. Make sure the configuration array has a section that says: `s:25:"payment_request_button_locations";a:2:{i:0;s:7:"product";i:1;s:4:"cart";}`.

### Test filter configuration persistence

1. Install and activate the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
2. Go to **Snippets > Add new**.
3. Copy the following snippet into the **Code** text area. Save this snippet as **Show PRBs on the Checkout Page**.

```php
// Snippet name: Show PRBs on the Checkout Page.
function show_prb_on_checkout_page($should_show) {
	return true;
}
add_filter( 'wc_stripe_show_payment_request_on_checkout', 'show_prb_on_checkout_page' );
```

4. Make sure **Run snippet everywhere** is toggled on.
5. Click **Save changes and activate**.
6. Repeat the snippet creation and activation for the following 2 snippets; make sure both are in separate snippets with distinct names:

```php
// Snippet name: Hide PRBs on Product Pages.
function hide_prb_on_product_page( $hide_prb ) {
	return true;
}
add_filter( 'wc_stripe_hide_payment_request_on_product_page', 'hide_prb_on_product_page' );
```

```php
// Snippet name: Hide PRBs on the Cart Page.
function hide_prb_on_cart_page( $should_show_prb ) {
	return false;
}
add_filter( 'wc_stripe_show_payment_request_on_cart', 'hide_prb_on_cart_page' );
```

7. You should now have 3 snippets active:

<img width="672" alt="image" src="https://user-images.githubusercontent.com/13835680/129649163-c827fbcf-d9f0-465a-820c-2cdc5e0b1db4.png">

8. Open phpMyAdmin, navigate to the **wp_options** table, toggle **Show all** on, search for **woocommerce_stripe_settings**, and click **Edit**.

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/13835680/129649524-a174d450-e508-460a-9ab7-6b7514a0b0a0.png">

9. Make the following changes to the **option_value**, and click **Go**:
	* Decrement the initial `a:XX` value by one. For example if the first letters are `a:24` change that to `a:23`.
	* Remove the whole `s:25:"payment_request_button_locations";a:2:{ ... }` part of the settings array. This will most likely be the last configuration in the settings array.

<details><summary>Example of changes applied to **option_value**</summary>
Change

```
a:24:{s:7:"enabled";s:3:"yes";s:5:"title";s:20:"Credit Card (Stripe)";s:11:"description";s:37:"Pay with your credit card via Stripe.";s:15:"api_credentials";s:0:"";s:8:"testmode";s:3:"yes";s:20:"test_publishable_key";s:42:"pk_test_xxxx";s:15:"test_secret_key";s:42:"sk_test_xxxx";s:15:"publishable_key";s:0:"";s:10:"secret_key";s:0:"";s:7:"webhook";s:0:"";s:19:"test_webhook_secret";s:38:"whsec_xxxx";s:14:"webhook_secret";s:0:"";s:14:"inline_cc_form";s:3:"yes";s:20:"statement_descriptor";s:7:"ATMTTC";s:7:"capture";s:3:"yes";s:15:"payment_request";s:3:"yes";s:27:"payment_request_button_type";s:3:"buy";s:28:"payment_request_button_theme";s:4:"dark";s:29:"payment_request_button_height";s:2:"44";s:28:"payment_request_button_label";s:7:"Buy now";s:35:"payment_request_button_branded_type";s:4:"long";s:11:"saved_cards";s:3:"yes";s:7:"logging";s:3:"yes";s:25:"payment_request_button_locations";a:2:{i:0;s:4:"cart";i:1;s:8:"checkout";}}
```

to

```
a:23:{s:7:"enabled";s:3:"yes";s:5:"title";s:20:"Credit Card (Stripe)";s:11:"description";s:37:"Pay with your credit card via Stripe.";s:15:"api_credentials";s:0:"";s:8:"testmode";s:3:"yes";s:20:"test_publishable_key";s:42:"pk_test_xxxx";s:15:"test_secret_key";s:42:"sk_test_xxxx";s:15:"publishable_key";s:0:"";s:10:"secret_key";s:0:"";s:7:"webhook";s:0:"";s:19:"test_webhook_secret";s:38:"whsec_xxxx";s:14:"webhook_secret";s:0:"";s:14:"inline_cc_form";s:3:"yes";s:20:"statement_descriptor";s:7:"ATMTTC";s:7:"capture";s:3:"yes";s:15:"payment_request";s:3:"yes";s:27:"payment_request_button_type";s:3:"buy";s:28:"payment_request_button_theme";s:4:"dark";s:29:"payment_request_button_height";s:2:"44";s:28:"payment_request_button_label";s:7:"Buy now";s:35:"payment_request_button_branded_type";s:4:"long";s:11:"saved_cards";s:3:"yes";s:7:"logging";s:3:"yes";}
```
</details>

10. Change the `WC_STRIPE_VERSION` constant [in `woocommerce-gateway-stripe.php:24`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/38744bb1c4add59869f5067734c0960f726b1228/woocommerce-gateway-stripe.php#L24) from whatever it's current value is to any other version, e.g. change `5.3.0` to `5.4.0`.
11. Open **Snippets** and disable any active snippets.
12. Open **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**
13. Make sure the **Payment Request Button Locations** option is set to **Checkout**.
14. Make sure the PRBs are loaded on the store's **Checkout** page (both Blocks and Shortcode versions).
15. Make sure the PRBs are not loaded on the store's **Cart** page (both Blocks and Shortcode versions).
16. Make sure the PRBs are not loaded on any of the store's **Product** pages.
17. Open phpMyAdmin, navigate to the **wp_options** table, toggle **Show all** on, search for **woocommerce_stripe_settings**, and click **Edit**.
18. Make sure the configuration array has a section that says: `s:25:"payment_request_button_locations";a:1:{i:0;s:8:"checkout";}`.

Now repeat steps 8 to 18 with all possible combinations of the snippets enabled/disabled:

| Show PRBs on the Checkout Page | Hide PRBs on Product Pages | Hide PRBs on the Cart Page | Configuration array value | PRB locations displayed |
|:---:|:---:|:---:|:---:|:---:|
| ✅ | ✅ | ✅ | `s:25:"payment_request_button_locations";a:1:{i:0;s:8:"checkout";}` | **Checkout** |
| ✅ | ✅ | ❌ | `s:25:"payment_request_button_locations";a:2:{i:0;s:4:"cart";i:1;s:8:"checkout";}` | **Cart**, **Checkout** |
| ✅ | ❌ | ✅ | `s:25:"payment_request_button_locations";a:2:{i:0;s:7:"product";i:1;s:8:"checkout";}` | **Product**, **Checkout** |
| ❌ | ✅ | ✅ | `s:25:"payment_request_button_locations";a:0:{}` | - |
| ✅ | ❌ | ❌ | `s:25:"payment_request_button_locations";a:3:{i:0;s:7:"product";i:1;s:4:"cart";i:2;s:8:"checkout";}` | **Product**, **Cart**, **Checkout** |
| ❌ | ✅ | ❌ | `s:25:"payment_request_button_locations";a:1:{i:0;s:4:"cart";}` | **Cart** |
| ❌ | ❌ | ✅ | `s:25:"payment_request_button_locations";a:1:{i:0;s:7:"product";}` | **Product** |
| ❌ | ❌ | ❌ | `s:25:"payment_request_button_locations";a:2:{i:0;s:7:"product";i:1;s:4:"cart";}` | **Product**, **Cart** |

### Test changing the settings

1. Open **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**
2. Choose any combination of **Product**, **Cart**, and **Checkout** as the value for **Payment Request Button Locations**.
3. Click **Save changes**
4. Make sure the Payment Request Buttons are only loaded on the pages you chose.

<img width="672" alt="image" src="https://user-images.githubusercontent.com/13835680/129644892-9627c448-6f0f-4b35-9b94-a6f1006e4271.png">

Repeat the testing instructions with all possible combinations of the PRB locations:

1. No locations selected.
1. **Product**.
2. **Cart**.
3. **Checkout**.
4. **Product**, **Cart**.
5. **Product**, **Checkout**.
6. **Cart**, **Checkout**.
7. **Product**, **Cart**, **Checkout**.

### Test filters taking precedence over plugin settings

1. Open **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**
2. Enable **Product** and **Cart** in the **Payment Request Button Locations** option.
3. Open **Snippets** and enable all the Code Snippets you added during one of the previous tests.
4. Make sure the PRBs are loaded on the store's **Checkout** page (both Blocks and Shortcode versions).
5. Make sure the PRBs are not loaded on the store's **Cart** page (both Blocks and Shortcode versions).
6. Make sure the PRBs are not loaded on any of the store's **Product** pages.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
add settings for PRB locations

Commit:
use settings instead of filters to hide/show PRBs

reykjalin commented:
I pinged @LevinMedia in Slack to get some feedback on the UX for this PR; just adding a comment here to let @asumaran (reviewer) know that I'd also like to see a review of the UX here from David 🙂 
It shouldn't delay the regular review process though.

asumaran commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -129,13 +129,14 @@ private function should_show_payment_request_button() {
 		//       to version 5.0.
 		if ( function_exists( 'has_block' ) ) {
-			global $post;
+			$prb_locations = empty( $this->settings['payment_request_locations'] )
+				? [] : $this->settings['payment_request_locations'];
```
According to the [Readme](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/readme.txt) the minimum version we support is `5.6` otherwise we would use the null coalescing operator (`??` ) here [available since PHP 7](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op), ending up like:

```php
$prb_locations = $this->settings['payment_request_locations'] ?? [];
```

Which makes me wonder... Officially, what's the minimum PHP version we support? The Readme says we support PHP 5.6 but then what's the minimum version of WooCommerce we currently support? Because if the minimum WooCommerce version we support is `3.9` then we should [bump the minimum version](https://docs.woocommerce.com/document/update-php-wordpress/) for this plugin to `7.0` right? Because there's no way to use this plugin without WooCommerce.

--

Based on the GitHub actions it seems though we support [WooCommerce 3.3.0 with PHP 5.6](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/cf24c32bd57bdb5cb8577da9bd500d34087eeb5b/.github/workflows/compatibility.yml#L18-L21) then nevermind 🤷‍♂️

reykjalin replied:
Good thought, but not relevant right now 🙂 

We're on track to get there though! We'll be bumping the minimum required versions in the coming months 🎉

asumaran commented on includes/admin/stripe-settings.php:
```diff
@@ -180,6 +180,18 @@
+				'cart'     => __( 'Cart page', 'woocommerce-gateway-stripe' ),
+				'checkout' => __( 'Checkout page', 'woocommerce-gateway-stripe' ),
+			],
+			'default'     => [ 'product', 'cart' ],
```
There's one issue here. This will not work until the user submits the Stripe settings form. The two options will be selected while the value is not persisted in the DB yet.

<img width="659" alt="Screen Shot 2021-08-11 at 12 04 44 AM" src="https://user-images.githubusercontent.com/1025173/128972477-5c3352de-827c-4fba-8295-e9e7fa7ee241.png">

Is this going to be saved/updated automatically when the user updates the plugin? Otherwise we will show incorrect state for this setting to the user.

reykjalin replied:
> This will not work until the user submits the Stripe settings form. The two options will be selected while the value is not persisted in the DB yet.

This will work in the sense that if there's nothing saved to the DB the default values are used, and that's intended. Nothing should (or has) to be saved to the database until the merchant updates the plugin settings.

I added some logic in e3ad0f5655125d5cc7b3ad6b9641a128ed2ba591 to retrieve the values during an update, but I don't think we should keep that logic. We're also discussing this in p1628641383050300-slack-CHG7MTCAF. I've pinged you in the thread.

reykjalin replied:
Ah, like we discussed in Slack, you were right; there was an issue with relying on the default value.

I've fixed that in e6901e7a30c5ed9500a3facf36f24711b3965d6c 🎉

asumaran replied:
I think e3ad0f5 also fixes the issue. But not for when switching to your branch and `payment_request_location` was not defined yet. This last case is fixed by e6901e7

asumaran reviewed (commented):
Tests well! I left a couple comments for you @reykjalin

Commit:
use woo enhanced multiselect

Commit:
update prb location settings using filter values during plugin update

Commit:
fix behavior when relying on default values

asumaran reviewed (approved):
LGTM! 🎉

Commit:
fix bug when prbs are hidden everywhere

reykjalin commented:
@asumaran I added one bug-fix in df5221e3ccf5406dffcc8bf5858f9dae65808276 and updated the testing instructions.

The testing instructions are a bit complicated to ensure the filter persistence code is functioning correctly, but the code for that section hasn't changed.

This should be the last version of this PR; if this is approved then we're good 🙌 


**EDIT:** I also rebased on `develop` and added a changelog entry 🙂

Commit:
add changelog entry

reykjalin commented:
@asumaran I just realized this PR isn't quite finished. We'll want to add deprecation warnings and still use the filters. I'll ping you again when it's ready for review. Sorry about the premature ping!

Commit:
add deprecation warnings to filters

Commit:
update changelog entry

reykjalin commented:
@asumaran This PR is now ready for a final review.

There have been some changes since last time:

1. Filters always take precedence over the plugin settings.
2. Using the filters results in deprecation warnings on product, cart, and checkout pages when the corresponding filter is being used.

asumaran commented:
Quick note: 

> Added the `payment_request_button_locations` setting.

You mean `payment_request_locations`? I think the above one is a better name

---

Also,

> Make sure the "Payment Request Locations" setting is set to Cart and Product.

Note label of the setting is "Payment Request Button Locations"

reykjalin commented:
> Quick note:
> 
> > Added the `payment_request_button_locations` setting.
> 
> You mean `payment_request_locations`? I think the above one is a better name

I agree with you, I'll change this. Thank you! 🙏 

> > Make sure the "Payment Request Locations" setting is set to Cart and Product.
> 
> Note label of the setting is "Payment Request Button Locations"

Thank you! I've fixed the testing instructions 🥇

Commit:
rename config option to payment_request_button_location

reykjalin commented:
@asumaran I renamed the config option to `payment_request_button_locations` in 63f8d50be04cb394e5096fb3e9d277320075c189.

asumaran commented:
@reykjalin when applying the steps for `Test filter configuration persistence` I've noticed the following:

> 11. Open Snippets and disable any active snippets.

If filters are activated or applied after the upgrade these are not used anymore. Even though the code says they are going to be respected.

```php
function hide_prb_on_product_page( $hide_prb ) {
	return false;
}
add_filter( 'wc_stripe_hide_payment_request_on_product_page', 'hide_prb_on_product_page' );

function hide_prb_on_cart_page( $should_show_prb ) {
	return false;
}
add_filter( 'wc_stripe_show_payment_request_on_cart', 'hide_prb_on_cart_page' );

function show_prb_on_checkout_page($should_show) {
	return true;
}
add_filter( 'wc_stripe_show_payment_request_on_checkout', 'show_prb_on_checkout_page' );
```

Using filters to show the button on all places doesn't update the dropdown in the Stripe settings form

![image](https://user-images.githubusercontent.com/1025173/129786582-d100d79a-f325-4950-b741-bcbd5cf009b1.png)

On this state the PRB is showing in the product and checkout pages only.

asumaran commented on woocommerce-gateway-stripe.php:
```diff
@@ -225,6 +225,52 @@ public function install() {
 					$this->update_plugin_version();
+
+					// TODO: Remove this when we're reasonably sure most merchants have had their
+					// settings updated like this. ~80% of merchants is a good threshold.
```
how do we know what amount of users have migrated to `5.4.0` or newer?

reykjalin replied:
We have some numbers on the spread of Stripe plugin versions in use that we can fall back on here 🙂

asumaran commented on woocommerce-gateway-stripe.php:
```diff
@@ -225,6 +225,52 @@ public function install() {
+				$prb_locations   = isset( $stripe_settings['payment_request_button_locations'] )
+					? $stripe_settings['payment_request_button_locations']
+					: [];
+				if ( ! empty( $stripe_settings ) && empty( $prb_locations ) ) {
```
```php
if ( empty( $prb_locations ) ) {
```

I think just cheking on `$prb_locations` is enough here?

reykjalin replied:
I'm not so sure 🤔 

If there are no plugin settings available, i.e. `$stripe_settings` is empty, then `empty( $prb_locations )` would evaluate to `true`, whereas `! empty( $stripe_settings ) && empty( $prb_locations )` would evaluate to `false`.

I'm not sure what implications it has to have the `$prb_locations` be the only thing written to the stripe settings, but I figured it'd be better to fall back on default values if there are no Stripe plugin settings available.

What do you think?

asumaran replied:
Thanks for the explanation. I figured out why you are checking on `$stripe_settings` too. You're right!

asumaran reviewed (approved):
Looks good! Tests well. Thanks for detailed testing steps! Left a couple non-blocking comments.
